### ISSUE TITLE
BERT PyTorch Wiki Dataset Download Error

### ISSUE THREAD

**Issue by roclark** [NVIDIA/DeepLearningExamples#840]
Related to **Model/Framework(s)** 
BERT PyTorch Model

**Describe the bug**
The Wikimedia data for BERT does not download as it can't verify the website's certificates. This throws the following stacktrace:

```
Resolving dumps.wikimedia.your.org (dumps.wikimedia.your.org)... 2001:4978:1:420::cc09:3752, 204.9.55.82
Connecting to dumps.wikimedia.your.org (dumps.wikimedia.your.org)|2001:4978:1:420::cc09:3752|:443... connected.
ERROR: cannot verify dumps.wikimedia.your.org's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:
  Unable to locally verify the issuer's authority.
To connect to dumps.wikimedia.your.org insecurely, use `--no-check-certificate'.
Traceback (most recent call last):
  File "/workspace/bert/data/bertPrep.py", line 363, in <module>
    main(args)
  File "/workspace/bert/data/bertPrep.py", line 60, in main
    downloader.download()
  File "/workspace/bert/data/Downloader.py", line 33, in download
    self.download_wikicorpus('en')
  File "/workspace/bert/data/Downloader.py", line 70, in download_wikicorpus
    downloader.download()
  File "/workspace/bert/data/WikiDownloader.py", line 54, in download
    raise RuntimeError('Wiki download not successful')
RuntimeError: Wiki download not successful
```

I haven't tested this out, but I assume that adding `--no-check-certificate` to [this line](https://github.com/NVIDIA/DeepLearningExamples/blob/3459f97d5884b2d3c629d6ddd3e8dcb2a6d1ab7e/PyTorch/LanguageModeling/BERT/data/WikiDownloader.py#L50) would allow the download to continue as expected.

**To Reproduce**
1. Follow the steps in the BERT PyTorch [README](https://github.com/NVIDIA/DeepLearningExamples/tree/3459f97/PyTorch/LanguageModeling/BERT) to setup the container.
2. Build the container using `build scripts/docker/build.sh`.
3. Run `/workspace/bert/data/create_datasets_from_start.sh`.
4. The error shown above will be displayed after a few seconds.

**Expected behavior**
I expect no errors to be thrown and the Wiki data to be downloaded and extracted.

**Environment**
Please provide at least:
* Container version: The built `bert` image following the [README](https://github.com/NVIDIA/DeepLearningExamples/tree/3459f97/PyTorch/LanguageModeling/BERT) (doesn't contain any tags or specifics).
* GPUs in the system: 2x Quadro RTX 8000 48GB
* CUDA driver version: 460.32.03

**Comment by roclark** [comment]
Update: Looks like the `--no-check-certificate` addition on the line mentioned above allows the download to occur, but this illuminates another issue where the previous attempt creates an empty file, so running the download again makes it _think_ the file exists, skipping the download, but throws some errors while trying to extract the empty file. Here's the output while running through this:

```
Downloading: https://dumps.wikimedia.your.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2
** Download file already exists, skipping download
Unzipping: wikicorpus_en.xml.bz2

bzip2: Compressed file ends unexpectedly;
	perhaps it is corrupted?  *Possible* reason follows.
bzip2: No such file or directory
	Input file = /workspace/bert/data/download/wikicorpus_en/wikicorpus_en.xml.bz2, output file = /workspace/bert/data/download/wikicorpus_en/wikicorpus_en.xml

It is possible that the compressed file(s) have become corrupted.
You can use the -tvv option to test integrity of such files.

You can use the `bzip2recover' program to attempt to recover
data from undamaged sections of corrupted files.

bzip2: Deleting output file /workspace/bert/data/download/wikicorpus_en/wikicorpus_en.xml, if it exists.
Traceback (most recent call last):
  File "/workspace/bert/data/bertPrep.py", line 363, in <module>
    main(args)
  File "/workspace/bert/data/bertPrep.py", line 60, in main
    downloader.download()
  File "/workspace/bert/data/Downloader.py", line 33, in download
    self.download_wikicorpus('en')
  File "/workspace/bert/data/Downloader.py", line 70, in download_wikicorpus
    downloader.download()
  File "/workspace/bert/data/WikiDownloader.py", line 58, in download
    subprocess.run('bzip2 -dk ' + self.save_path + '/' + filename, shell=True, check=True)
  File "/opt/conda/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'bzip2 -dk /workspace/bert/data/download/wikicorpus_en/wikicorpus_en.xml.bz2' returned non-zero exit status 2.
```

**Comment by sharathts** [comment]
Correct, if the previous download isn't fully complete (there is no way of programatically checking the download's completeness), the extraction on the partially downloaded wikicorpus_en.xml.bz2 fails.

The best solution is to manually remove corrupt/incomplete downloads and redo it again.

**Comment by roclark** [comment]
While I agree that there's no way to tell if the download completed programatically since it doesn't appear there are any checksums to compare with on the Wiki website, it is still possible to see if a file is empty in Python which would be the most likely scenario encountered I would assume. I was able to delete the file and resume downloading (using the `--no-check-certificate` flag mentioned), but perhaps some documentation in the README or handling in the code would help other users? Thanks for the help, BTW.

**Comment by sharathts** [comment]
Thanks for the suggestion. Will reflect the necessary info in the documentation with our next update.

**Comment by jzhang82119** [comment]
adding --no-check-certificate, but I still cannot get wiki download to work


/workspace/bert# /workspace/bert/data/create_datasets_from_start.sh --no-check-certificate
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
Working Directory: /workspace/bert/data
Action: download
Dataset Name: wikicorpus_en

Directory Structure:
{ 'download': '/workspace/bert/data/download',
  'extracted': '/workspace/bert/data/extracted',
  'formatted': '/workspace/bert/data/formatted_one_article_per_line',
  'hdf5': '/workspace/bert/data/hdf5_lower_case_1_seq_len_512_max_pred_20_masked_lm_prob_0.15_random_seed_12345_dupe_factor_5',
  'sharded': '/workspace/bert/data/sharded_training_shards_256_test_shards_256_fraction_0.1',
  'tfrecord': '/workspace/bert/data/tfrecord_lower_case_1_seq_len_512_max_pred_20_masked_lm_prob_0.15_random_seed_12345_dupe_factor_5'}

Downloading: https://dumps.wikimedia.your.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2
Running: ['wget', 'https://dumps.wikimedia.your.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2', '--output-document=/workspace/bert/data/download/wikicorpus_en/wikicorpus_en.xml.bz2']
--2021-02-24 19:57:54--  https://dumps.wikimedia.your.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2
Resolving dumps.wikimedia.your.org (dumps.wikimedia.your.org)... 204.9.55.82, 2001:4978:1:420::cc09:3752
Connecting to dumps.wikimedia.your.org (dumps.wikimedia.your.org)|204.9.55.82|:443... connected.
ERROR: cannot verify dumps.wikimedia.your.org's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:
  Unable to locally verify the issuer's authority.
To connect to dumps.wikimedia.your.org insecurely, use `--no-check-certificate'.
Traceback (most recent call last):
  File "/workspace/bert/data/bertPrep.py", line 363, in <module>
    main(args)
  File "/workspace/bert/data/bertPrep.py", line 60, in main
    downloader.download()
  File "/workspace/bert/data/Downloader.py", line 33, in download
    self.download_wikicorpus('en')
  File "/workspace/bert/data/Downloader.py", line 70, in download_wikicorpus
    downloader.download()
  File "/workspace/bert/data/WikiDownloader.py", line 54, in download
    raise RuntimeError('Wiki download not successful')
RuntimeError: Wiki download not successful

**Comment by roclark** [comment]
Hey @jzhang82119, how did you implement the `--no-check-certificate` flag? Looks like from your output that it wasn't actually added to the code. If it helps, this is what I changed my line listed above to:
```python
cmd = ['wget', url, '--no-check-certificate', '--output-document={}'.format(self.save_path + '/' + filename)]
```

**Comment by jzhang82119** [comment]
@roclark , I thought --no-check-certificate flag were to added after /workspace/bert/data/create_datasets_from_start.sh, I just read it carefully, I will modify data/WikiDownloader.py#L50

**Comment by jzhang82119** [comment]
Yes. it works. Thanks.

**Comment by roclark** [comment]
Glad that worked out for you, thanks!

**Comment by roclark** [comment]
Looks like this is fixed with #841. Thanks!
